pkg/fileutils: GetTotalUsedFds: reduce allocations

thaJeztah opened the pull request:
- relates to https://github.com/moby/moby/issues/45842

### pkg/fileutils: GetTotalUsedFds(): don't pretend to support FreeBSD

Commit 8d56108ffb4e334600377c4bb4471eecec7b825c moved this function from
the generic (no build-tags) fileutils.go to a unix file, adding "freebsd"
to the build-tags.

This likely was a wrong assumption (as other files had freebsd build-tags).
FreeBSD's procfs does not mention `/proc/<pid>/fd` in the manpage, and
we don't test FreeBSD in CI, so let's drop it, and make this a Linux-only
file.

While updating also dropping the import-tag, as we're planning to move
this file internal to the daemon.


### pkg/fileutils: add BenchmarkGetTotalUsedFds

    go test -bench ^BenchmarkGetTotalUsedFds$ -run ^$ ./pkg/fileutils/
    goos: linux
    goarch: arm64
    pkg: github.com/docker/docker/pkg/fileutils
    BenchmarkGetTotalUsedFds-5   	  149272	      7896 ns/op	     945 B/op	      20 allocs/op

### pkg/fileutils: GetTotalUsedFds: reduce allocations

Use File.Readdirnames instead of os.ReadDir, as we're only interested in
the number of files, and results don't have to be sorted.

Before:

    BenchmarkGetTotalUsedFds-5            149272              7896 ns/op             945 B/op         20 allocs/op

After:

    BenchmarkGetTotalUsedFds-5            153517              7644 ns/op             408 B/op         10 allocs/op


### pkg/fileutils: GetTotalUsedFds(): use fast-path for Kernel 6.2 and up

Linux 6.2 and up (commit [f1f1f2569901ec5b9d425f2e91c09a0e320768f3][1])
provides a fast path for the number of open files for the process.

From the [Linux docs][2]:

> The number of open files for the process is stored in 'size' member of
> `stat()` output for /proc/<pid>/fd for fast access.

[1]: https://github.com/torvalds/linux/commit/f1f1f2569901ec5b9d425f2e91c09a0e320768f3
[2]: https://docs.kernel.org/filesystems/proc.html#proc-pid-fd-list-of-symlinks-to-open-files

This patch adds a fast-path for Kernels that support this, and falls back
to the slow path if the Size fields is zero.

Comparing on a Fedora 38 (kernel 6.2.9-300.fc38.x86_64):

Before/After:

    go test -bench ^BenchmarkGetTotalUsedFds$ -run ^$ ./pkg/fileutils/
    BenchmarkGetTotalUsedFds        57264     18595 ns/op     408 B/op      10 allocs/op
    BenchmarkGetTotalUsedFds       370392      3271 ns/op      40 B/op       3 allocs/op

Note that the slow path has 1 more file-descriptor, due to the open
file-handle for /proc/<pid>/fd during the calculation.

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
pkg/fileutils: add BenchmarkGetTotalUsedFds

    go test -bench ^BenchmarkGetTotalUsedFds$ -run ^$ ./pkg/fileutils/
    goos: linux
    goarch: arm64
    pkg: github.com/docker/docker/pkg/fileutils
    BenchmarkGetTotalUsedFds-5   	  149272	      7896 ns/op	     945 B/op	      20 allocs/op

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/fileutils: GetTotalUsedFds: reduce allocations

Use File.Readdirnames instead of os.ReadDir, as we're only interested in
the number of files, and results don't have to be sorted.

Before:

    BenchmarkGetTotalUsedFds-5   	  149272	      7896 ns/op	     945 B/op	      20 allocs/op

After:

    BenchmarkGetTotalUsedFds-5   	  153517	      7644 ns/op	     408 B/op	      10 allocs/op

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on pkg/fileutils/fileutils_unix.go:
```diff
@@ -13,10 +13,18 @@ import (
-	return -1
+	defer f.Close()
+
+	dirs, err := f.Readdirnames(-1)
```
This should not use `-1`, and we can read it in a loop until EOF

akerouanton commented on pkg/fileutils/fileutils_unix.go:
```diff
@@ -6,17 +6,93 @@ import (
-		log.G(context.TODO()).Errorf("Error opening /proc/%d/fd: %s", os.Getpid(), err)
-	} else {
-		return len(fds)
+	nameCount, err := readDirnamesCount(fmt.Sprintf("/proc/%d/fd", os.Getpid()))
```
From https://docs.kernel.org/filesystems/proc.html#proc-pid-fd-list-of-symlinks-to-open-files:

> The number of open files for the process is stored in 'size' member of stat() output for /proc/<pid>/fd for fast access.

neersighted replied:
Per @corhere; only on kernel 6.4+. We should use this as a fast path, and fall back to `readdir(2)`

corhere replied:
For future reference: the fast path was added to the kernel in https://github.com/torvalds/linux/commit/f1f1f2569901ec5b9d425f2e91c09a0e320768f3

thaJeztah committed:
pkg/fileutils: GetTotalUsedFds(): use fast-path for Kernel 6.2 and up

Linux 6.2 and up (commit [f1f1f2569901ec5b9d425f2e91c09a0e320768f3][1])
provides a fast path for the number of open files for the process.

From the [Linux docs][2]:

> The number of open files for the process is stored in 'size' member of
> `stat()` output for /proc/<pid>/fd for fast access.

[1]: https://github.com/torvalds/linux/commit/f1f1f2569901ec5b9d425f2e91c09a0e320768f3
[2]: https://docs.kernel.org/filesystems/proc.html#proc-pid-fd-list-of-symlinks-to-open-files

This patch adds a fast-path for Kernels that support this, and falls back
to the slow path if the Size fields is zero.

Comparing on a Fedora 38 (kernel 6.2.9-300.fc38.x86_64):

Before/After:

    go test -bench ^BenchmarkGetTotalUsedFds$ -run ^$ ./pkg/fileutils/
    BenchmarkGetTotalUsedFds        57264     18595 ns/op     408 B/op      10 allocs/op
    BenchmarkGetTotalUsedFds       370392      3271 ns/op      40 B/op       3 allocs/op

Note that the slow path has 1 more file-descriptor, due to the open
file-handle for /proc/<pid>/fd during the calculation.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

corhere commented on pkg/fileutils/fileutils_unix.go:
```diff
@@ -5,18 +5,51 @@ package fileutils // import "github.com/docker/docker/pkg/fileutils"
+	}
+	defer f.Close()
+
+	var (
+		fdCount int
+		names   = make([]string, 0, 100)
+	)
+	for {
+		names, err = f.Readdirnames(100)
```
Pre-allocating a slice for `names` is a waste as `f.Readdirnames` returns a new slice.
```suggestion
	var fdCount int
	for {
		names, err := f.Readdirnames(100)
```

thaJeztah replied:
Doh! Good call. I had it because initially I calculated `size(names)` outside the for-loop, then realised that wouldn't work of course 😂

thaJeztah committed:
pkg/fileutils: GetTotalUsedFds(): don't pretend to support FreeBSD

Commit 8d56108ffb4e334600377c4bb4471eecec7b825c moved this function from
the generic (no build-tags) fileutils.go to a unix file, adding "freebsd"
to the build-tags.

This likely was a wrong assumption (as other files had freebsd build-tags).
FreeBSD's procfs does not mention `/proc/<pid>/fd` in the manpage, and
we don't test FreeBSD in CI, so let's drop it, and make this a Linux-only
file.

While updating also dropping the import-tag, as we're planning to move
this file internal to the daemon.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on pkg/fileutils/fileutils_unix.go:
I had a look, and it looks like the freebsd build-tag was added in https://github.com/moby/moby/commit/8d56108ffb4e334600377c4bb4471eecec7b825c (https://github.com/moby/moby/pull/16407), which very likely was "because other files had that build-tag as well".

I pushed an extra commit to make it Linux-only.

thaJeztah commented on pkg/fileutils/fileutils_linux.go:
```diff
@@ -0,0 +1,52 @@
+			return -1
+		}
+	}
+	// Note that the slow path has 1 more file-descriptor, due to the open
+	// file-handle for /proc/<pid>/fd during the calculation.
+	return fdCount
```
This was a fun one; "wait, why is there an off-by-one?" ahhhh... wait... gotcha.
